Home_work17/Player.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+#include <string>
+#include<iostream>
+#include "PlayerClass.h"
+#include "Weapon.h"
+
+
+class Player
+{
+public:
+	Player(const std::string& name = "Auto Generated NameId", Weapon* weapon = new BareHand, PlayerClass* playerClass = new DefaultClass)
+		: m_name(name), m_weapon(weapon), m_class(playerClass) {
+		setHealth(100 * playerClass->getHealthMultiplier());
+	}
+	~Player()
+	{
+		delete m_class;
+		delete m_weapon;
+	}
+
+
+	std::string getName() const { return m_name; }
+	int getBaseDamage(int distance) { return m_weapon->getDamageRate(distance); }
+
+	void setWeapon(Weapon* weapon)
+	{
+		delete m_weapon;
+		m_weapon = weapon;
+	}
+	void setPlayerClass(PlayerClass* playerClass) {
+		{
+			delete m_class;
+			m_class = playerClass;
+		}
+	}
+	int getHealth() { return m_health; }
+	void setHealth(int health) { m_health = health; }
+	bool isEliminated() const { return m_health <= 0; }
+
+	void attack(Player& other, int distance)
+	{
+		if (&other == this)
+		{
+			std::cout << "\n-------FRIENDLY FIRE!-------\n";
+		}
+
+		if (other.isEliminated())
+		{
+			std::cout << "Player " << other.getName() << " has been already eliminated!\n";
+			return;
+		}
+
+		//#TODO: Optional: Add your own damage-related modifier
+		float multipler = 0.f;
+		if (dynamic_cast<RangedWeapon *>(m_weapon) != nullptr) {
+			//std::cout << " Ranged attack \n";
+			multipler = m_class->getRangedDamageMultiplier();
+		}
+		else if (dynamic_cast<MeleeWeapon*>(m_weapon) != nullptr) {
+			//std::cout << " Melee attack \n";
+			multipler = m_class->getMeleeDamageMultiplier();
+		}
```

Comment:
Weapon class should have getDamageMultiplier() method and you can override it for all derieved classes. So here you don't need cast anything, just call `multiplier = m_weapon->getDamageMultiplier();`